keel-core/src/main/kotlin/com/netflix/spinnaker/keel/actuation/EnvironmentPromotionChecker.kt

```diff
@@ -141,14 +157,21 @@ class EnvironmentPromotionChecker(
                    * We don't need to re-invoke stateful constraint evaluators for these, but we still
                    * check stateless constraints to avoid approval outside of allowed-times.
                    */
+                  log.info("version: [$v] of artifact ${artifact.name} is in queuedForApproval for environment ${environment.name}")
                   if (checkStatelessConstraints(artifact, deliveryConfig, v, environment)) {
                     approveVersion(deliveryConfig, artifact, v, environment.name)
                     repository.deleteQueuedConstraintApproval(deliveryConfig.name, environment.name, v)
                   }
                 }
             }
             if (!approvedPending && versionIsPending || version == null) {
-              log.warn("No version of {} passes constraints for environment {}", artifact.name, environment.name)
```

Comment:
This warning was not so true. The condition here does not cover the case where an older version gets approved (lets sat version 1 is approved and version 2 is pending, so: `approvedPending` is false (since version 1 is not pending approval anymore), `versionIsPending` is true (for version 2 which is pending), and vesion==2).

Reply:
You description here seems to match exactly what the `if` statement expresses. Unless there's an operator precedence issue, where you might rewrite it as:
```kotlin
if ((!approvedPending && versionIsPending) || version == null) {
```
In your example:
```kotlin
if ((!false && true) || false)
```
...which would evaluate to `true` and enter this block, right?

Reply:
Never mind. You mean you _do_ enter the block as expected, it's just the warning was in the wrong place. Makes sense.